app/src/heartbeat/heartbeat_storage_desktop.cc

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "app/src/heartbeat/heartbeat_storage_desktop.h"
+
+#include <fstream>
+#include <regex>
+#include <vector>
+
+#include "app/logged_heartbeats_generated.h"
+#include "app/src/filesystem.h"
+#include "app/src/include/firebase/internal/mutex.h"
+
+namespace firebase {
+namespace heartbeat {
+
+using LoggedHeartbeatsFbs =
+    com::google::firebase::cpp::heartbeat::LoggedHeartbeats;
+using com::google::firebase::cpp::heartbeat::CreateLoggedHeartbeats;
+using com::google::firebase::cpp::heartbeat::CreateUserAgentAndDates;
+using com::google::firebase::cpp::heartbeat::GetLoggedHeartbeats;
+using com::google::firebase::cpp::heartbeat::UserAgentAndDates;
+using com::google::firebase::cpp::heartbeat::VerifyLoggedHeartbeatsBuffer;
+
+// Using an anonymous namespace for file mutex and filename
+namespace {
+
+const char kHeartbeatDir[] = "firebase-heartbeat";
+const char kHeartbeatFilenamePrefix[] = "heartbeats-";
+
+// Returns the mutex that protects accesses to the storage file.
+Mutex& FileMutex() {
```

Comment:
I think this mutex is better to be in class level.

That is, heartbeat storages for app1 and app2 are not necessarily writing to the same location. BUT it is also possible that they are write to the same location as well. Rare but possible.

Beside, static mutex is a sources of troubles. Since you did not write anything to destroy it, that is memory leak as well.

On the other hand, if you can guarantee that the access to this API are all serialized to the same thread, like using `schedular` class, the mutex is actually not required.

Reply:
Ok. For now I removed the mutex.
Deciding whether to rely on some new mutex, or scheduler, or just OS file locking is something that I think can be added while working on the controller.